### PR TITLE
Improve --auto transformations

### LINKED ISSUE #40: Incorrect migration output when using colon-delimiting and invoke
> The generated/converted code on the critical line is:
`await wrapZafClient(this.zafClient, "ticketFields");`

> ...when it should have maybe been something like:
`await wrapZafClient(this.zafClient, "ticketFields:sharedWith.hide");`

> However, even the second call above doesn't work (I tried).

### PR DESCRIPTION
**Pull request by svizzari**
:v:

/cc @zendesk/vegemite

### Description
Fixes:
- Correctly form path when using `invoke` and intermediate arguments
- That a bound statement was being replaced when it shouldn't be, e.g. `this.when().then(this.foo.bind(this))` was being transformed to `this.when().then(await this.foo.bind(this))` (which just won't work 🙀)

Resolves #40

New transforms:
- `this.when` to `Promise.all`
- `this.promise` to `new Promise`
- `"abc%@".fmt(1)` to `helpers.fmt("abc%@", 1)` 

### References
* #40 

### Risks
* [low] Only affects specific migration paths

### CODE REVIEW

File: src/plugins/sync_to_async.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
 import * as types from "babel-types";
-import { compact, uniq, chain, get } from "lodash";
+import { compact, uniq, chain, get, last } from "lodash";
```
Comment (hoangtrvu) [inline src/plugins/sync_to_async.ts]: is `last` needed here?

File: src/plugins/sync_to_async.ts [hunk 2]
```diff
@@ -214,7 +216,7 @@ const syncToAsyncVisitor = {
       // If `name` appears in `apis`, it will be a call to a v1 api.
       // Next, try work out whether it is a get, set, or invoke (depending
```
Comment (hoangtrvu) [inline src/plugins/sync_to_async.ts]: this comment is obsolete now

File: src/plugins/sync_to_async.ts [hunk 3]
```diff
@@ -214,7 +216,7 @@ const syncToAsyncVisitor = {
...
       // on whether there were any arguments passed to the exp).
-      if (exp.node.arguments.length) {
+      if (/^(ticket|user|organization)Fields$/.test(name)) {
```
Comment (hoangtrvu) [inline src/plugins/sync_to_async.ts]: as discussed, `getExpressionToReplace` is already called at this point

File: src/test/steps/migrate_app_js.test.ts [hunk 4]
```diff
@@ -308,6 +308,17 @@ describe("migrate app js", () => {
                      }`)
                   );
                 });
+                it("should correctly chain together invoke calls", async () => {
+                  writeFixtureSrc(`foo: function() {
+                    this.${api}("sharedWith").hide();
```
Comment (hoangtrvu) [inline src/test/steps/migrate_app_js.test.ts]: just a note for you to double check what kind of apis and arguments can be chained, what about multiple arguments

File: src/test/steps/migrate_app_js.test.ts [hunk 5]
```diff
@@ -308,6 +308,17 @@ describe("migrate app js", () => {
...
+                   }`);
+                  await subject(options);
+                  expect(readMigratedSrc()).to.have.string(
+                    wrapExpectedSrc(`foo: async function() {
+                       await wrapZafClient(this.zafClient, "${api}:sharedWith.hide");
```
Comment (hoangtrvu) [inline src/test/steps/migrate_app_js.test.ts]: note for you to double check if this line works properly (see https://zendesk.slack.com/files/U1EHCH8PM/F7AMGF58S/disablesharingdemo.zip)

File: src/plugins/sync_to_async.ts (hunk 1 above)
Comment (svizzari) [reply src/plugins/sync_to_async.ts]: Nope, I refactored out the use of `last`, will clean up

**Pull request merged**